File: abc259/d.cpp
#include <bits/stdc++.h>
#include <cmath>
#define ll long long
using namespace std;

int main()
{
    int N;
    cin >> N;
    ll sx, sy, tx, ty;
    cin >> sx >> sy >> tx >> ty;
    vector<vector<ll>> Circles(N);
    for (int i = 0; i < N; i++)
    {
        ll x, y, r;
        cin >> x >> y >> r;
        Circles[i].push_back(x);
        Circles[i].push_back(y);
        Circles[i].push_back(r);
    }

    // calculate start circle & end circle;

    int s = -1;
    int t = -1;
    for (int i = 0; i < N; i++)
    {
        auto ci = Circles[i];
        auto xi = ci[0];
        auto yi = ci[1];
        auto ri = ci[2];
        double distance_s = sqrt(pow(xi - sx, 2) + pow(yi - sy, 2));
        if (distance_s == double(ri))
        {
            s = i;
        }
        double distance_t = sqrt(pow(xi - tx, 2) + pow(yi - ty, 2));
        if (distance_t == double(ri))
        {
            t = i;
        }
    }

    // cout << s << ", " << t << endl;

    // Create Graph
    vector<vector<int>> Graph(N);
    for (int i = 0; i < N - 1; i++)
    {
        for (int j = i + 1; j < N; j++)
        {
            auto ci = Circles[i];
            auto cj = Circles[j];
            // Check whether ci & cj are connected.
            auto xi = ci[0];
            auto yi = ci[1];
            auto ri = ci[2];
            auto xj = cj[0];
            auto yj = cj[1];
            auto rj = cj[2];
            // ll distance_between_centers = pow(xi - xj, 2) + pow(yi - yj, 2);
            ll distance_between_centers = (xi - xj) * (xi - xj) + (yi - yj) * (yi - yj);
            bool is_connected = !(distance_between_centers > (ri+rj)*(ri+rj) || distance_between_centers < (ri-rj)*(ri-rj));
            // cout << is_connected << endl;
            if (is_connected)
            {
                Graph[i].push_back(j);
                Graph[j].push_back(i);
            }
        }
    }

    // BFS

    queue<int> q;
    q.push(s);
    vector<bool> visited(N);
    visited[s] = true;
    while (!q.empty() && !visited[t])
    {
        auto cur = q.front();
        q.pop();
        for (auto next : Graph[cur])
        {
            if (!visited[next])
            {
                q.push(next);
                visited[next] = true;
            }
        }
    }
    if (visited[t])
    {
        cout << "Yes" << endl;
    }
    else
    {
        cout << "No" << endl;
    }
}

File: abc259/d.py
from collections import deque
N = int(input())
sx,sy,tx,ty = map(int,input().split())
circles = [
    list(map(int,input().split())) for _ in range(N)
]


graph = [[] for _ in range(N)]

for i in  range(N-1):
    for j in range(i+1,N):
        ci,cj = circles[i],circles[j]
        xi,yi,ri = ci
        xj,yj,rj = cj
        d = (xi-xj)**2 + (yi-yj) ** 2
        if (ri - rj) ** 2 <= d <= (ri+rj) ** 2:
            graph[i].append(j)
            graph[j].append(i)

start = -1
end = -1
for i in range(N):
    xi,yi,ri = circles[i]
    if (sx - xi) ** 2 + (sy-yi) ** 2 == ri ** 2:
        start = i
    if (tx - xi) ** 2 + (ty-yi) ** 2 == ri ** 2:
        end = i

# BFS

q = deque([start])
visited = [False for _ in range(N)]
visited[start] = True

while len(q)>0 and not visited[end]:
    cur = q.popleft()
    for n in graph[cur]:
        if not visited[n]:
            q.append(n)
            visited[n] = True
            if visited[end]:
                break


if visited[end]:
    print("Yes")
else:
    print("No")